### PR TITLE
Add minimal forwarding RPC server for host driven python execution on Hexagon

### PR DESCRIPTION
**Pull request by csullivan**
This PR adds a proxy RPC server for Hexagon. It is a wrapper which takes standard TVM RPC calls from a python host to a remote Android device and forwards them across FastRPC to Hexagon. This RPC flow will be replaced by running a minimal RPC server directly on Hexagon. For now we provide a prototype forwarding RPC server for host driven execution on Hexagon.

This PR also implements a few supporting fixes:
* Adds `HexagonHostModuleNode` which is now the default runtime hexagon runtime module. It separates out the current functionality of HexagonModule from the FastRPC offload device api in `src/runtime/hexagon/android`. This allows `libtvm.so` to be built with codegen support for hexagon, without target or simulator runtime support. 
* Updates CodeGenHexagon to sync. with recent changes to the LLVM code generator
    * A TODO from the PR is to consider if we can move Hexagon over to using the LLVM runtime module. In that case `module.export_library` could provide DSO linking rather than codegen. 
* Add CodeGenHexagon::AddMainFunction to expose `__tvm_main__`.

### CODE REVIEW

File: apps/hexagon_proxy_rpc/rpc_env.cc [hunk 1]
```diff
@@ -0,0 +1,321 @@
+/*
```
Comment (kparzysz-quic) [inline apps/hexagon_proxy_rpc/rpc_env.cc:1]: How is this loaded?  Would it make sense to have this under src/runtime/hexagon/rpc?

File: src/runtime/hexagon/rpc/device_api.cc [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+  void StreamSync(Device dev, TVMStreamHandle stream) final {}
+  void* AllocWorkspace(Device dev, size_t size, DLDataType type_hint) final{
+    throw tvm::runtime::Error("HexagonRPCDeviceAPI::AllocWorkspace is unimplemented");
+  };
+  void FreeWorkspace(Device dev, void* data) final {
+    throw tvm::runtime::Error("HexagonRPCDeviceAPI::FreeWorkspace is unimplemented");
+  };
+
+  static HexagonRPCDeviceAPI* Global() {
+    static auto* inst = new HexagonRPCDeviceAPI();
```
[the remark this answers is not in the document]
Comment (csullivan) [reply src/runtime/hexagon/rpc/device_api.cc]: Here I just followed the [convention](https://github.com/apache/tvm/blob/main/src/runtime/cpu_device_api.cc#L77) of the other device APIs on their global singleton access.

File: apps/hexagon_proxy_rpc/rpc_env.cc (hunk 1 above)
Comment (csullivan) [reply apps/hexagon_proxy_rpc/rpc_env.cc:1]: `rpc_env.cc` is a hexagon version of [rpc_env.cc](https://github.com/apache/tvm/blob/main/apps/cpp_rpc/rpc_env.cc) from apps/cpp_rpc.  In this case, all other sources from the C++ RPC server are reused.  As noted in the description above we are working on a solution to run a minrpc server directly on hexagon which will provide more general functionality (e.g. support for tuning). This minrpc implementation will land in `src/runtime/hexagon/rpc as you suggest`. Until then, we are submitting this prototype forwarding RPC server application for use in development.

Comment (csullivan) [reply apps/hexagon_proxy_rpc/rpc_env.cc:1]: cc @mehrdadh who is working on the minrpc server implementation.

Comment (mehrdadh) [reply apps/hexagon_proxy_rpc/rpc_env.cc:1]: I agree with @csullivan. Let's keep these files under `app` and he minRPC server under `src`

File: cmake/config.cmake [hunk 3]
```diff
@@ -285,6 +285,9 @@ set(USE_HEXAGON_SDK /path/to/sdk)
 # Whether to build the hexagon launcher
 set(USE_HEXAGON_LAUNCHER OFF)
 
+# Whether to build the minimal support android rpc server for hexagon
+set(USE_MINIMAL_HEXAGON_TVM_RPC OFF)
```
Comment (mehrdadh) [inline cmake/config.cmake:289]: Maybe make this more explicit? `USE_HEXAGON_PROXY_RPC`?

Comment (mehrdadh) [reply cmake/config.cmake:289]: also you could add it to tvm/CMakeList.txt as a `tvm_option` with some description.

Comment (csullivan) [reply cmake/config.cmake]: Great point, I've added these changes in [8d0e7d7](https://github.com/apache/tvm/pull/9526/commits/8d0e7d7faa30bd1c1a748561e796fa704b7fb7ed).

**Review by jwfromm — APPROVED** [review]
LGTM, thanks Chris!

**Comment by TejashShah** [comment]
Thanks, Chris! 🙌

**Pull request merged**